### PR TITLE
prometheus-operator: Make sure all ServiceMonitors are selected using label

### PR DESCRIPTION
**Pull request by surajssd**
This commit sets the value of `serviceMonitorSelectorNilUsesHelmValues`
to `false` to make sure that all `ServiceMonitors` are selected.

When the value of `serviceMonitorSelectorNilUsesHelmValues` is `true`
the `Prometheus` object's field `serviceMonitorSelector` is set with
following values:

```yaml
serviceMonitorSelector:
  matchLabels:
    release: "prometheus-operator"
```

Due to this any `ServiceMonitor` object that does not have above labels,
they won't be selected. For Lokomotive we would want to keep track of
all ServiceMonitors in the cluster.

And when `serviceMonitorSelectorNilUsesHelmValues` is set to `false`,
`serviceMonitorSelector` field in `Prometheus` object is empty.

Fixes https://github.com/kinvolk/lokomotive/issues/19

### CODE REVIEW

**Comment by surajssd** [comment]
Why increase work for some unforeseen situation? When we face such a situation where our prom operator cannot handle all the service monitors in the cluster then we can add a knob to fix the problem. 

I don't want to do premature optimization, that's all.

**Comment by invidian** [comment]
>Why increase work for some unforeseen situation? When we face such a situation where our prom operator cannot handle all the service monitors in the cluster then we can add a knob to fix the problem.

You cannot control what ServiceMonitors user create, so I think it's sane to do the separation, to make sure cluster (system) monitoring is separated from the user workload monitoring.

**Comment by surajssd** [comment]
> You cannot control what ServiceMonitors user create, so I think it's sane to do the separation, to make sure cluster (system) monitoring is separated from the user workload monitoring.

I think this is premature optimization and we are not in a situation where prometheus cannot manage the targets it should scrape. I think such sharding should come only after we encounter such a situation to keep things simple.

**Comment by iaguis** [comment]
I agree with @invidian here, I think it would be surprising and unexpected for the user if we'd track all the their ServiceMonitors objects by default and it makes sense that we track only ours.

Maybe @rata has some more experience/opinions here?

**Comment by rata** [comment]
I don't have any strong opinions, so feel free to ignore me. Some thoughts, though.

I think we are guessing quite a lot here. And maybe we can have a default and a configuration knob to let the users that wants to set `serviceMonitorSelectorNilUsesHelmValues` change it. If it is not that obvious what value should be fixed, maybe it should be tunnable? :)

I'm not sure we have a nice option, though. If we pick all ServiceMonitor, do we have the flexibility for the user to add grafana charts, etc. for user workloads? (that not overridden on upgrades, etc.). And if we don't, the user needs to monitor user workloads using a different grafana (or maybe even prometheus) instance, then? Doesn't seem like the best UX, IMHO (having two different monitoring solutions is really a bad option, specially if it complicates to correlate system components data and user workloads data to understand outages, etc.).

Maybe going for the latter (let the user use the same prometheus instance explicitelly if the want, by adding labels, and deploy their own grafana they can customize) is the best path for now? And add a knob so they can set `serviceMonitorSelectorNilUsesHelmValues` to pick all ServiceMonitor if they want too?

Does it make sense?

**Comment by surajssd** [comment]
Made changes to the PR to include a variable which user can tweak, but the default behavior is to select all the `ServiceMonitor` in the cluster.

If going forward we get lot of user requirement where users want to have different prometheus for cluster and different for apps then we can think of changing the default.

Also if you look at the contemporary monitoring scenario where people are trying to converge prometheus instances into one not shard them. Projects like Thanos and Cortex show that want to have a single pane of window to query data from.

**Comment by invidian** [comment]
@surajssd thanks for adding the knob. I agree with @rata that this is right solution. However, I think it should be `true` by default and it should still collect all cluster ServiceMonitor objects. If user wants to opt-in and use single Prometheus instance for everything, it should be their choice, not ours.

**Comment by surajssd** [comment]
> However, I think it should be true by default and it should still collect all cluster ServiceMonitor objects. If user wants to opt-in and use single Prometheus instance for everything, it should be their choice, not ours.

I think it should be our decision as to what should be default and IMHO the default to set it to false is the correct one. As with other the way we are opinionated I would argue that we do the same with this decision as well. And if users complain I am more than happy to revisit this decision.

**Comment by surajssd** [comment]
@rata @iaguis @invidian what is should we do to move this PR forward?

**Comment by invidian** [comment]
@surajssd I'd do two things to get it resolved:
- add appropriate labels for all ServiceMonitor objects we deploy, so we monitor them all by default
- add a knob to gather all ServiceMonitor objects, which should be **opt-in**

**Comment by surajssd** [comment]
But then that is the point of disagreement.

**Comment by iaguis** [comment]
I think that if we monitor everything in the cluster by default some users would like it and think that's nice and some would be surprised.

I don't know what kind of user would be the most common and I understand @surajssd's point of being opinionated but following the [Principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) I'd lean towards not monitoring everything in the cluster by default so I'd agree with @invidian.

We can document that if users want workloads to be monitored by l8e's Prometheus they should add the right label.

On an unrelated note I also think the label should be different since `release: "prometheus-operator"` sounds confusing to me. Something like `prometheus: lokomotive` or so.

**Comment by surajssd** [comment]
@iaguis made those changes.

> On an unrelated note I also think the label should be different since `release: "prometheus-operator"` sounds confusing to me. Something like `prometheus: lokomotive` or so.

This is possible without changing the proemetheus chart itself. Right now this name is auto generated and is equal to the release name. And we have our release name set to proemetheus-operator!

File: docs/configuration-reference/components/prometheus-operator.md [hunk 1]
```diff
@@ -91,6 +91,8 @@ Example:
 | `prometheus_metrics_retention` | Time duration Prometheus shall retain data for. Must match the regular expression `[0-9]+(ms\|s\|m\|h\|d\|w\|y)` (milliseconds, seconds, minutes, hours, days, weeks and years). | `10d` | false |
 | `prometheus_external_url` | The external URL Prometheus instances will be available under. This is necessary to generate correct URLs. This is necessary if Prometheus is not served from root of a DNS name. | "" | false |
 | `prometheus_node_selector` | Node selector to specify nodes where the Prometheus pods should be deployed. | {} | false |
+| `watch_labeled_service_monitors` | By default prometheus operator watches all the ServiceMonitor objects in the cluster. If set to `true` then only ServiceMonitors with label `release: "prometheus-operator"` will be watched. | `true` | false |
```
Comment (invidian) [inline docs/configuration-reference/components/prometheus-operator.md]: The description of parameter is not correct as far as I understand. We agreed not to collect all of them by default.

File: docs/configuration-reference/components/prometheus-operator.md [hunk 2]
```diff
@@ -91,6 +91,8 @@ Example:
...
+| `watch_labeled_prometheus_rules` | By default prometheus operator watches all the PrometheusRule objects in the cluster. If set to `true` then only PrometheusRule with label `release: "prometheus-operator"` will be watched. | `true` | false |
```
Comment (invidian) [inline docs/configuration-reference/components/prometheus-operator.md]: Ditto.

File: docs/configuration-reference/components/prometheus-operator.md (hunk 1 above)
Comment (surajssd) [reply docs/configuration-reference/components/prometheus-operator.md]: good catch. I forgot to change this unlike other things.

**Comment by surajssd** [comment]
@iaguis 

> We should follow up with changing all components to set those labels, right?

Yes that the next thing to do on my plate.

**Pull request merged**